clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerNetworkClient.java

```diff
@@ -256,11 +256,13 @@ public void poll(long timeout, long now, PollCondition pollCondition) {
      * nor will it execute any delayed tasks.
      */
     public void pollNoWakeup() {
-        disableWakeups();
-        try {
-            poll(0, time.milliseconds(), null);
-        } finally {
-            enableWakeups();
+        synchronized (this) {
```

Comment:
@ijuma the problem here is that I need to lock onto something for the sequence from `disableWakeups();` and `enableWakeups();`, otherwise I'll run into trouble when the heartbeat thread calls this method, disabling the wakeups, while the test simultaneously calls `maybeTriggerWakeup`, which then might run into `wakeupDisabledCount != 0`.
That would be the issue solved here, but I see the issue with `fireCompletedPendingRequests` then running in a context synchronized on `this`. Maybe this could just be solved by sync on a different lock just for this case? Synchronize on `this` is not needed I think.